src/phase-management/PhaseAdvancer.js

```diff
@@ -0,0 +1,249 @@
+          detail: `Rule ${rule.name} failed`,
+          failureReasons: ruleExecutionResult.failureReasons,
+        };
+      }
+    }
+
+    if (operation === "open") {
+      await this.#open(challengeId, phases, phase);
+    } else if (operation === "close") {
+      await this.#close(challengeId, phases, phase);
+    }
+
+    const next = operation === "close" ? phases.filter((p) => p.predecessor === phase.phaseId) : [];
+
+    return {
+      success: true,
+      message: `Successfully ${operation}d phase ${phase.name} for challenge ${challengeId}`,
+      updatedPhases: phases,
+      next,
+    };
+  }
+
+  async #open(challengeId, phases, phase) {
+    console.log(`Opening phase ${phase.name} for challenge ${challengeId}`);
+
+    phase.isOpen = true;
+    const actualStartDate = new Date();
+    phase.actualStartDate = actualStartDate.toISOString();
+    phase.scheduledEndDate = new Date(
+      actualStartDate.getTime() + phase.duration * 1000
+    ).toISOString();
+
+    const scheduledStartDate = parseDate(phase.scheduledStartDate);
+    const delta = scheduledStartDate - actualStartDate; // in milliseconds
+
+    if (delta !== 0) {
+      this.#updateSubsequentPhases(phases, phase, -delta);
+    }
+
+    console.log(`Updated phases: ${JSON.stringify(phases, null, 2)}`);
+  }
+
+  async #close(challengeId, phases, phase) {
+    console.log(`Closing phase ${phase.name} for challenge ${challengeId}`);
+
+    phase.isOpen = false;
+    const actualEndDate = new Date();
+    phase.actualEndDate = actualEndDate.toISOString();
+
+    const scheduledEndDate = parseDate(phase.scheduledEndDate);
+    const delta = scheduledEndDate - actualEndDate;
+
+    if (delta !== 0) {
+      this.#updateSubsequentPhases(phases, phase, -delta);
+    }
+
+    console.log(`Updated phases: ${JSON.stringify(phases, null, 2)}`);
+  }
+
+  #isPastTime(dateString) {
+    if (dateString == null) {
+      return false;
+    }
+    const date = new Date(dateString);
+    const now = new Date();
+    return date <= now;
+  }
+
+  async #getRegistrantCount(challengeId) {
+    console.log(`Getting registrant count for challenge ${challengeId}`);
+    // TODO: getChallengeResources loops through all pages, which is not necessary here, we can just get the first page and total count from header[X-Total]
+    const submitters = await helper.getChallengeResources(challengeId, config.SUBMITTER_ROLE_ID);
+    return submitters.length;
+  }
+
+  async #getSubmissionCount(challengeId) {
+    console.log(`Getting submission count for challenge ${challengeId}`);
+    // TODO: getChallengeSubmissions loops through all pages, which is not necessary here, we can just get the first page and total count from header[X-Total]
+    const submissions = await helper.getChallengeSubmissions(challengeId);
+    return submissions.length;
+  }
+
+  async #areAllSubmissionsReviewed(challengeId) {
+    console.log(`Getting review count for challenge ${challengeId}`);
```

Comment:
Note: also we need to check if enough review is present per submission according to "Reviewer Count" phase constraint.

Reply:
I will double check - however areAllSubmissionReviewed is an `essential rule` while checking if all submissions received the minimum requisite reviews would be a `constraint check`